lib/Conversion/TorchToMhlo/BasicOp.cpp

```diff
@@ -125,180 +127,219 @@ class ConvertAtenConstPatternOp : public OpConversionPattern<AtenOpT> {
 
 } // namespace
 
-// These binary op legalizations are specific to add/sub which have an
+// These binary op legalizations are specific to add/sub(Tensor) which have an
 // alpha multiplier.
+// aten.add(sub).Tensor
 namespace {
 template <typename AtenOpT, typename MhloOpT>
-class ConvertAtenAddSubOp : public OpConversionPattern<AtenOpT> {
+class ConvertAtenAddSubTensorOp : public OpConversionPattern<AtenOpT> {
 public:
   using OpConversionPattern<AtenOpT>::OpConversionPattern;
   using OpAdaptor = typename AtenOpT::Adaptor;
   LogicalResult
   matchAndRewrite(AtenOpT op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
     Value lhs = adaptor.self();
-    TensorType lhsType = lhs.getType().dyn_cast<TensorType>();
+    RankedTensorType lhsType = lhs.getType().dyn_cast<RankedTensorType>();
     Value rhs = adaptor.other();
-    TensorType rhsType = rhs.getType().dyn_cast<TensorType>();
+    RankedTensorType rhsType = rhs.getType().dyn_cast<RankedTensorType>();
 
-    if (!lhsType)
+    if (!lhsType || !rhsType)
       return op.emitError("Only Tensor types supported in MHLO");
 
-    TensorType outType = OpConversionPattern<AtenOpT>::getTypeConverter()
-                             ->convertType(op.getType())
-                             .template cast<TensorType>();
+    RankedTensorType outType = OpConversionPattern<AtenOpT>::getTypeConverter()
+                                   ->convertType(op.getType())
+                                   .template cast<RankedTensorType>();
 
     Type outElemTy = outType.getElementType();
     if (!outElemTy.isIntOrFloat()) {
       return op.emitError(
           "Only floating-point or integer datatype legalization supported");
     }
 
-    Value rhsAsTensor;
-    if (!rhsType) {
-      if (failed(mhlo::torchScalarToMhloTensor(rewriter, op, op.other(),
-                                               rhsAsTensor, outElemTy,
-                                               outType.getShape())))
-        return op.emitError("Currently only scalar constants are supported for "
-                            "conversion in MHLO operation");
-    }
-    Value lhsTensor = lhs;
-    Value rhsTensor = rhsType ? rhs : rhsAsTensor;
+    Value bcastShape;
 
     // Handle broadcasting. Since we have the output type already, here we
     // just broodcast operands' shape to output shape.
-    lhsTensor = mhlo::promoteAndBroadcast(rewriter, lhsTensor, outType);
-    rhsTensor = mhlo::promoteAndBroadcast(rewriter, rhsTensor, outType);
+    if (lhsType.hasStaticShape() && rhsType.hasStaticShape()) {
```

Comment:
why do we need broadcastBinaryOperandsDynamic separate from promoteAndBroadcast? In the static case shouldn't they canonicalize to the same thing?

Reply:
Yeah, we have tried to handle static shape as a special case of dynamic. But supporting dynamic shape will import shape ops and more nested blocks. We are worrying that it probably does harm to the efficiency when lowering static shaped ops.

Reply:
Perhaps we can use `mhlo::DimOp` && `tensor::FromElements` instead of `shape::Dialect`?

Reply:
For example, this is the `AtenBroadcastToOp` currently implemented in BladeDISC:
```
template <>
LogicalResult ConvertAtenOp<AtenBroadcastToOp>::matchAndRewrite(
    AtenBroadcastToOp op,
    OpAdaptor adaptor,
    ConversionPatternRewriter& rewriter) const {
  // Not a tensor type.
  //
  auto selfType = adaptor.self().getType().dyn_cast<TensorType>();
  if (!selfType)
    return op.emitError("Only tensor types are currently supported");

  SmallVector<Value> dimsSize;
  if (!getListConstructElements(adaptor.size(), dimsSize)) {
    return op.emitError("Dims size must be a list of Scalar");
  }

  auto loc = op.getLoc();
  auto rankType = selfType.dyn_cast<RankedTensorType>();
  auto selfRank = rankType ? rankType.getRank() : 0;
  auto newRank = dimsSize.size();
  auto leadingRank = newRank - selfRank;
  for (size_t d = 0; d < newRank; ++d) {
    // !torch.int
    auto dsize = dimsSize[d];
    int64_t dval;
    if (matchPattern(dsize, m_TorchConstantInt(&dval)) && dval == -1) {
      if (d < leadingRank) {
        return op.emitError(
            "Ref: https://pytorch.org/docs/stable/generated/torch.Tensor.expand.html."
            "For the new leading dimensions, the size cannot be set to -1.");
      } else {
        // Passing -1 as the size for a dimension means not changing the size
        // of that dimension.
        //
        // tensor.dim %self -> index
        dsize = rewriter.create<tensor::DimOp>(
            loc, adaptor.self(), d - leadingRank);
      }
    } else {
      // !torch.int -> i64
      dsize = rewriter.create<ToI64Op>(loc, dsize).getResult();
      // i64 -> index
      dsize = rewriter.create<mlir::arith::IndexCastOp>(
          loc, rewriter.getIndexType(), dsize);
    }
    // index -> i32
    dsize = rewriter.create<mlir::arith::IndexCastOp>(
        loc, rewriter.getI32Type(), dsize);
    dimsSize[d] = dsize;
  }

  auto mhloShape = rewriter.create<mlir::tensor::FromElementsOp>(loc, dimsSize);
  auto broadcastDims =
      BuildI64ElementsAttr(rewriter, mhlo::rangeIndices(leadingRank, newRank));
  rewriter.replaceOpWithNewOp<mhlo::DynamicBroadcastInDimOp>(
      op,
      getTypeConverter()->convertType(op.getType()),
      adaptor.self(),
      mhloShape,
      broadcastDims);
  return success();
}
```

Reply:
Another bonus is that we cat avoid introducing shape dialect.

Reply:
As we want to check whether two dynamic shaped tensor is broadcastable, shape ops is easier to use than `tensor.DimOp` & `tensor.FromElementsOp`

Reply:
Then I would prefer `chlo::BroadcastBinOp` then shape dialect

Reply:
Since our target is MHLO dialect, introducing CHLO dialect probably needs an additional pipeline to lowering CHLO to MHLO. We are trying to avoid this.

Reply:
> why do we need broadcastBinaryOperandsDynamic separate from promoteAndBroadcast? In the static case shouldn't they canonicalize to the same thing?
> Yes, I think a better way to support static shapes is to do canonicalization in some passes for static shapes. And static shapes from ShapeType should not be used by default.

Thanks a lot. We will unify the implementation of static shaped and dynamic shaped ops' lowering. And let the canonicalizer to eliminate unnecessary shape ops when involved tensors are totally static shaped.

Reply:
This is a very good discussion, if you discover some ergonomic/representation/abstraction issues here please let us know. These op sets are not set in stone (some more difficult to change, but not infeasible :-)) and feedback very important to being able to improve them.

Reply:
@fortianyou @Vremold Introducing `chlo` is ok for me, (which might ease our development for dynamic cases), as long as we add a chlo_legalize_to_hlo transform in our lowering pipeline. `ShapeDialect` should be allowed in necessary situation (if chlo is not strong enough), since the `chlo_legalize_to_hlo` also introduces it. Does this sound good to you guys?

Reply:
> @fortianyou @Vremold Introducing `chlo` is ok for me, (which might ease our development for dynamic cases), as long as we add the `chlo_legalize_to_hlo` transform in our lowering pipeline. `ShapeDialect` should be allowed in necessary situation (if chlo is not strong enough), since the `chlo_legalize_to_hlo` also introduces it. Does this sound good to you guys?

Okay for me.

Reply:
It seems that an agreement is reached. I will close this PR and switch to use CHLO dialect in `convert-torch-to-mhlo` pass to handle dynamic cases. While in the whole `torch-backend-to-mhlo-backend-pipeline` pipeline, we plan to reuse `chlo_legalize_to_hlo` pass in mlir-hlo community to finally convert CHLO ops to MHLO ops, which might introduce a little change in current CMake files. What do you think of this plan? @silvasean @ZihengJiang @fortianyou

Reply:
Sounds good. Thanks for the effort :)

Reply:
Sounds good! Great discussion!

Reply:
Great! Also, I think we can make chlo_legalize_to_hlo configurable :)